AI 챗봇 앱 [STEP 4] 벨로, 밤삭

JinUng41 opened the pull request:
리뷰어: @forestjae

### 팀원

| 프로필 사진 | <a href="https://github.com/bamsak"><img src="https://avatars.githubusercontent.com/u/114239407?v=4" width="90" height="90"></a> | <a href="https://github.com/JinUng41"><img src="https://avatars.githubusercontent.com/u/91656206?v=4" width=90></a> | 
| ---- | ---------- | --------- | 
| in Github | [@bamsak](https://github.com/bamsak) | [@JinUng41](https://github.com/JinUng41) 
| in SeSAC | 밤삭 | 벨로

안녕하세요, 숲재!!

잘 부탁드립니다!! 🙇‍♂️

[STEP 4]에서 CoreData를 활용하여 채팅 내역을 저장하고,

채팅 목록들을 리스트로 보여줄 수 있도록 구현해 보았습니다.

### 핵심 요구사항

- [x]  Safe Area을 고려한 오토레이아웃 구현
- [x]  CoreData를 활용한 데이터 저장 및 불러오기
- [x]  Date 타입과 DateFormatter의 이해
- [x]  CoreData의 데이터 삭제 구현

## 구현 내용

### 폴더 구조

```
.
├── Application
│   ├── AppDelegate.swift
│   └── SceneDelegate.swift
├── Chat
│   ├── Controllers
│   │   └── GPTChatViewController.swift
│   ├── ViewModels
│   │   └── GPTChatViewModel.swift
│   └── Views
│       ├── ChatBubbleView.swift
│       ├── MessageCell.swift
│       ├── MessageContentConfiguration.swift
│       └── MessageContentView.swift
├── Extensions
│   ├── Bundle+Extension.swift
│   ├── Date+Extension.swift
│   ├── GPTChatRoomViewController+Extension.swift
│   ├── JSONDecoder+Extension.swift
│   ├── JSONEncoder+Extension.swift
│   ├── UITextView+Extension.swift
│   └── URLSession+Extension.swift
├── Models
│   ├── API
│   │   ├── APIEndPoint
│   │   │   ├── Implementation
│   │   │   │   └── GPTEndPoint.swift
│   │   │   └── Interface
│   │   │       └── APIEndPoint.swift
│   │   ├── Error
│   │   │   └── APIError.swift
│   │   └── HTTPMethod.swift
│   ├── CoreData
│   │   ├── DataModel.xcdatamodeld
│   │   │   └── DataModel.xcdatamodel
│   │   │       └── contents
│   │   ├── MessageEntity+CoreDataClass.swift
│   │   ├── MessageEntity+CoreDataProperties.swift
│   │   ├── RoomEntity+CoreDataClass.swift
│   │   └── RoomEntity+CoreDataProperties.swift
│   ├── DTOs
│   │   ├── Implementation
│   │   │   ├── Common
│   │   │   │   └── GPTMessageDTO.swift
│   │   │   ├── GPTRequestDTO.swift
│   │   │   └── GPTResponseDTO.swift
│   │   └── Interface
│   │       └── RequestEncodable+RequestDecodable.swift
│   └── Messages
│       ├── Implementation
│       │   ├── AssistantMessage.swift
│       │   ├── SystemMessage.swift
│       │   └── UserMessage.swift
│       └── Interface
│           └── GPTMessageable.swift
├── Protocols
│   ├── DataDecodable.swift
│   ├── DataEncodable.swift
│   └── URLSessionProtocol.swift
├── Resources
│   ├── Assets.xcassets
│   │   ├── AccentColor.colorset
│   │   │   └── Contents.json
│   │   ├── AppIcon.appiconset
│   │   │   └── Contents.json
│   │   └── Contents.json
│   ├── Base.lproj
│   │   └── LaunchScreen.storyboard
│   └── Info.plist
├── RoomList
│   ├── Controllers
│   │   └── GPTRoomListViewController.swift
│   └── ViewModel
│       └── GPTRoomListViewModel.swift
└── Utilities
    ├── CoreData
    │   ├── CoreDataManager
    │   │   ├── Implementation
    │   │   │   └── CoreDataManager.swift
    │   │   └── Interface
    │   │       └── DataManagable.swift
    │   └── DataHandler
    │       ├── ChatRoomDataHandler.swift
    │       ├── MessageDataHandler.swift
    │       └── Model
    │           ├── ChatMessage.swift
    │           └── ChatRoom.swift
    ├── Network
    │   └── NetworkManager.swift
    └── ServiceProvider
        ├── Implementation
        │   └── ServiceProvider.swift
        └── Interface
            └── ServiceProvidable.swift
```

### 동작화면

| 채팅방 생성 | CoreData 저장 | 최근 대화순 | 채팅방 삭제|
|-----------|----------|-----------|-----------|
|![채팅방 생성](https://github.com/tasty-code/ios-chat-bot/assets/91656206/20ac62b3-4f85-4998-8be8-e6ccc8ef64e9)|![CoreData 저장](https://github.com/tasty-code/ios-chat-bot/assets/91656206/fc35d7cb-1a0a-4178-9bc9-e4bbd469aa55)|![최근 대화순](https://github.com/tasty-code/ios-chat-bot/assets/91656206/3c304828-41c3-4e66-becf-3fcceaa8f9c1)|![삭제후재실행](https://github.com/tasty-code/ios-chat-bot/assets/91656206/028789c4-486b-44e3-a947-fcfe79fcf36f)


### 객체의 역할

- CoreDataManager
    - 싱글톤으로 구현되어 있으며, context와 NSPersistentContainer를 내부 프로퍼티로 가지며, context를 통해 현재 컨테이너 상태를 저장해주는 역할을 합니다.
- ChatRoomDataHandler
    - CoreDataManager를 통해 CoreData에 저장된 채팅방(RoomEntity) 리스트 데이터를 불러와 View에 보여질 ChatRoom으로 변환하여 반환해주는 역할을 해줍니다.
- MessageDataHandler
    - CoreDataManager를 통해 CoreData에 저장된 채팅(MessageEntity)데이터를 불러와 View에 보여질  ChatMessage로 변환해주는 역할을 해줍니다.
- ChatRoom
    - CoreData에 쓰이는 RoomEntity를 뷰에 보여주기 위해 사용되는 struct 타입입니다.
- ChatMessage
    - CoreData에 쓰이는 MessageEntity와 API 통신을 통해 얻을 수 있는 GPTMessageDTO를 뷰에 보여주기 위해 사용되는 struct 타입입니다.

### GPTMessageDTO → ChatMessage

STEP 3부터 STEP 4 중반까지 저희는 채팅을 보여주기 위해 GPTMessageDTO를 사용하고 있었습니다.

DiffableDataSource를 구성하기 위해, GPTMessageDTO에 uuid 프로퍼티를 추가하여 사용하였습니다.

```swift
struct GPTMessageDTO: Codable {
    let uuid: UUID
    
		// 생략..

    init(uuid: UUID = UUID(), role: MessageRole, content: String? = nil, name: String? = nil, toolCalls: [ToolCalls]? = nil) {
        self.uuid = uuid
        self.role = role
        self.content = content
        self.name = name
        self.toolCalls = toolCalls
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.uuid = UUID()
        self.role = try container.decode(MessageRole.self, forKey: .role)
        self.content = try container.decode(String.self, forKey: .content)
        self.name = try container.decodeIfPresent(String.self, forKey: .name)
        self.toolCalls = try container.decodeIfPresent([ToolCalls].self, forKey: .toolCalls)
    }
}
```

CoreData를 구현하면서 추가적으로 Date와 관련한 프로퍼티가 필요하게 되었습니다.

CoreData는 Relationship (Many - One)으로 저장되는 객체에 대하여 NSSet으로 저장합니다.

Set은 배열과 달리 순서를 보장하지 않기 때문에 sort를 위한 기준점이 필요하였고, 저희는 해당 기준점을 메세지가 생성되는 시점, 즉 Date로 바라보았습니다.

하지만 API 통신을 위한 (encoding, decoding) GPTMessageDTO가 Date타입을 추가하여 뷰에 보여주는 수준까지 쓰이는 것이 DTO로써의 역할에 맞지 않는다는 생각이 들었습니다.

따라서 GPTMessageDTO는 API 통신만을 위한 모델로 쓰이며, 해당 내용을 뷰에 보여주기 위해 ChatMessage를 추가로 구현하게 되었습니다.

```swift
struct ChatMessage {
    let uuid: UUID
    let date: Date
    var content: String?
    var role: MessageRole
    
    init(uuid: UUID = UUID(), date: Date = Date(), content: String?, role: MessageRole) {
        self.uuid = uuid
        self.date = date
        self.content = content
        self.role = role
    }
    
    func toMessageDTO() -> GPTMessageDTO {
        return GPTMessageDTO(role: role, content: content)
    }
}

extension ChatMessage: Hashable {
    static func == (lhs: Self, rhs: Self) -> Bool {
        return lhs.uuid == rhs.uuid
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(self.uuid)
    }
}
```

이렇게 함으로서 GPTMessageDTO는 API 통신 단에서, ChatMessage는 Present 단에서 사용하는 각각의 entity로 존재하게 되었습니다.

- 고민 되는 점
    - Clean Architecture에 대해서 정확히 알지는 못하지만 저희가 알고 있는 부분까지에 대해서 이와 같은 구현 방식이 Clean Architecture에서 설명하는 레이어 간의 엔터티 분리와 유사한지 여쭤보고 싶습니다.

### CoreDataManager를 싱글톤으로 구현한 이유

해당 객체를 싱글톤으로 구현하지 않았을 시, 가장 최상위 단부터 의존성 주입을 해주어야 하는데, 현재는 아직 해당 객체를 사용하는 부분이 런타임 시점에 다른 객체로 주입이 될 가능성이 없다고 판단이 듭니다.

그렇지만 NetworkManager에서 URLSession을 주입받는 형식처럼 후에 변경가능성을 고려하여 CoreDataManager를 프로토콜로 추상화 하여 의존성 주입을 통해 사용하도록 했습니다.

- 고민 되는 점
    - 현재 드는 고민 점은 CoreDataManager가 싱글톤으로 되어 있기 때문에 데이터를 관리하는 데에 thread safety하지 않아 문제가 발생할 수도 있을 거라는 생각이 듭니다.
    해당 부분을 Semaphore같은 thread 접근 제한을 거는 방식을 생각을 해보았는데, 아직은 어떤 시점에 Semaphore를 걸어줘야할 지 감이 잘 안옵니다. 해당 부분에 대해 조언을 얻고 싶습니다!

bamsak committed:
Merge pull request #3 from bamsak/step3

Step3 merge

JinUng41 committed:
[Chore] 코드 컨벤션에 맞게 코드 수정

- dataSource를 클로저로 구성
- 제네릭 타입 명시
- 공백 제거
- guard문의 else문 개행
- 빈줄 제거
- 파라미터명 변경

bamsak committed:
[Feat] ChatListView 초기 구성

bamsak committed:
[Feat] CoreDataManager 구현

- CoreData에 저장하기 위한 기초 객체 구현
- GPTMessageDTO 디코드 방식 변경

bamsak committed:
[Feat] Core Data Model 추가

- entity 정의
- 각 ViewController의 dataSource 정의 방식 변경 (강한 참조를 방지하기 위한 cell Registration을 dataSource 클로저 내부의 지역변수로 변경)
- ViewController와 ViewModel들 이름 수정

bamsak committed:
[Feat] 채팅방 목록 화면을 위한 ViewModel 구성

- Date Extension 추가
- ChatRoom 모델 추가 및 해당 모델로 diffable DataSource 구성
- RoomEntity를 ChatRoom으로 변환하는 Data handler 객체 구현

bamsak committed:
[Refactor] collectionView에 보여질 Model 변경

- GPTMessageDTO와 View간 분리
- ChatMessage 모델 구현(View에 보여질 Data)

bamsak committed:
[Feat] CoreData를 적용한 GPT챗 기능 구현

- GPT와의 채팅 내역을 CoreData로 저장해주는 역할을 하는 MessageDataHandler 구현 및 뷰 모델, 뷰 컨트롤러에서 연동

JinUng41 committed:
[Feat] 코어데이터 삭제 구현

- CoreData의 데이터 삭제를 위한 기능 구현
- 빈 줄 삭제
- dateFormat 수정

JinUng41 committed:
[Fix] 실시간으로 데이터의 시간이 반영되지 않는 문제 해결

forestjae commented on ChatBot/ChatBot/Chat/Controllers/GPTChatViewController.swift:
```diff
@@ -66,22 +66,24 @@ final class GPTChatRoomViewController: UIViewController {
-    private let cellRegistration: UICollectionView.CellRegistration<MessageCell, GPTMessageDTO> = {
-        return UICollectionView.CellRegistration<MessageCell, GPTMessageDTO> {
+    private lazy var dataSource: UICollectionViewDiffableDataSource<Section, ChatMessage> = {
+        let cellRegistration = UICollectionView.CellRegistration<MessageCell, ChatMessage> {
```
셀 Registration을 합치신 이유가 있나요? 가독성이 좀 안좋아진 느낌입니다.

bamsak replied:
기존 방식대로 했을 경우 dataSource 초기화 클로저 내부에서 cellRegistration 프로퍼티를 self 통해 참조를 하고 있었습니다.
이로 인해 화면을 빠져나옴에도 클로저 내부에 self에 대한 순환참조가 발생하여,
두번째 ViewController가  deinit이 되지 않는 현상을 발견함으로 메모리 누수현상이 발생하였습니다.
이를 해소하기 위해 위와 같은 방식으로 클로저 내부에서 지역변수로 cellRegistration을 선언하여 사용했습니다.

forestjae replied:
아 그렇군요. dataSource를 직접 변수 선언부에서 만들기 보다 만들어주는 메서드 내부에서 cellRegstration도 같이 만들면 더 보기 좋을 것 같네요

forestjae commented on ChatBot/ChatBot/Chat/Controllers/GPTChatViewController.swift:
```diff
@@ -100,17 +102,27 @@ final class GPTChatRoomViewController: UIViewController {
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        viewModel.saveMessages()
```
저장 시점을 메시지가 생성되자마자로 바꾸는건 어떨까요?

지금과 같은 상황에서는 특정 경우에 저장이 안 될 수도 있겠네요

JinUng41 replied:
네 말씀하신대로 저장 시점을 메세지가 생성되는 시점 즉, GPTChatViewModel의 `addResponseMessage(_:)`시점에서 저장하도록 수정하였습니다.

forestjae commented on ChatBot/ChatBot/Chat/Controllers/GPTChatViewController.swift:
```diff
@@ -156,18 +168,8 @@ final class GPTChatRoomViewController: UIViewController {
-    private func configureSnapshot(with messages: [GPTMessageDTO]) {
-        var snapshot = NSDiffableDataSourceSnapshot<Section, GPTMessageDTO>()
+    private func configureSnapshot(with messages: [ChatMessage]) {
+        var snapshot = NSDiffableDataSourceSnapshot<Section, ChatMessage>()
```
이런 스냅샷이나 디퍼블데이터소스의 경우 typealias를 활용하시면 좋습니다.

JinUng41 replied:
미처 생각해보지 못했던 부분이었습니다. 감사합니다. 🙏🙏

forestjae commented on ChatBot/ChatBot/Chat/ViewModels/GPTChatViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+    
+    var didMessagesAppend: (([ChatMessage]) -> Void)?
```
internal 한 프로퍼티는 항상 위쪽에 두시는 습관을 들이시는게 좋습니다.

JinUng41 replied:
네 수정하였습니다! 감사합니다!!👍👍

forestjae commented on ChatBot/ChatBot/Chat/ViewModels/GPTChatViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+        self.currentRoom = currentRoom
+    }
+    
+    func fetch(userInput: String) {
```
메서드도 마찬가지로 접근제어자에 따라서 순서 지켜주세요!

JinUng41 replied:
네 수정하였습니다!

forestjae commented on ChatBot/ChatBot/RoomList/Controllers/GPTRoomListViewController.swift:
```diff
@@ -0,0 +1,165 @@
+        else {
+            return nil
+        }
+        let deleteAction = UIContextualAction(style: .destructive, title: "삭제") {
+            [weak self] _, _, completion in
+            self?.viewModel.deleteChatRoom(at: indexPath.row)
+            completion(true)
```
💯 💯 💯

JinUng41 replied:
감사합니다!!

forestjae commented on ChatBot/ChatBot/Utilities/CoreData/CoreDataManager/Implementation/CoreDataManager.swift:
```diff
@@ -0,0 +1,35 @@
+}
+
+extension CoreDataManager: DataManagable {
+    var context: NSManagedObjectContext { container.viewContext }
```
viewContext가 어떤 특성을 가지는지 공부 해 보셨나요?

bamsak replied:
viewContext의 경우 mainThread에서 동작한다는 것을 알게되었습니다.
만약 무거운 작업이라면 mainThread에서 동작하는 viewContext보단 NSManagedObjectContext를 Concurrency타입으로 새로운 context를 만들거나 newBackgroundContext()을 이용하여 context를 생성해, 비동기로 동작을 시켜야 한다는 점을 알게 되었습니다.
이로 인해 PR로 질문 드렸던 부분은 제가 잘못생각하고 있었던 것 같습니다! 감사합니다!

forestjae commented on ChatBot/ChatBot/Utilities/CoreData/DataHandler/ChatRoomDataHandler.swift:
```diff
@@ -0,0 +1,51 @@
+
+import Foundation
+
+final class ChatRoomDataHandler {
```
👍

JinUng41 replied:
감사합니다!!

forestjae reviewed (approved):
정말 고생 많으셨습니다. 복잡한 뷰와, 데이터를 조작하는 로직들을 많이 작성하셔야 하는 프로젝트 였는데 정말 잘 해주신 것 같습니다. 네이밍과 컨벤션도 꼼꼼히 신경쓰신게 눈에 보입니다. 

질문하신 내용 답변드릴게요. 
> DTO의 사용
네 맞습니다. 용도에 잘 맞게 사용해 주셨어요. 불필요한 의존성을 없에거나, 불필요한 프로퍼티를 덜어냅니다.

> Singleton Coredata
우려하신 문제가 발생하는 이유는 NSManagedObject나 context들이 생성되고 사용되는 작업들이 여러 쓰레드에서 동시에 접근하는 경우일거에요. 지금 구현에서는 viewContext를 사용하셨기 때문에 항상 메인쓰레드에서 작업이 이루어질거고 이 경우에는 항상 메인쓰레드에서 사용하게 된다면 안전을 보장 할 수 있어요. 

만약 백그라운드에서 작업을 하는 기능을 가진 코어데이터 관리 객체를 만든다면, 그때 그때 해당하는 쓰레드에서 새로운 NSManagedObjectContext를 만들어내는 식으로 설계를 해야겠죠.

세마포어는 좀 더 고차원적인 문제를 해결할때 사용되는 개념이고, 이와 같이 낮은 단계에서의 문제 해결에 사용되기에는 어렵습니다.

NSManagedObjectContext에 대해서 좀 더 공부를 해보시면 더 명확해지실 거에요!

JinUng41 committed:
[Refactor] STEP 4 리뷰 반영

- 접근제어자에 따른 프로퍼티와 메서드 선언 순서 변경
- 메세지가 생성되는 시점에 CoreData에 저장하도록 변경
- CellRegistration, DiffableDataSource, Snapshot을 typealias로 선언
